x/pairing/keeper/msg_server_relay_payment.go

```diff
@@ -360,3 +362,61 @@ func appendRelayPaymentDetailsToEvent(from map[string]string, uniqueIdentifier u
 	}
 	return to
 }
+
+func (k Keeper) checkBadge(ctx sdk.Context, badgeData BadgeData, client string, relay *types.RelaySession) (newTimerExpiry uint64, err error) {
+	if !badgeData.Badge.IsBadgeValid(client, relay.LavaChainId, uint64(relay.Epoch)) {
+		return 0, utils.LavaFormatWarning("badge must match traits in relay request", fmt.Errorf("invalid badge"),
+			utils.Attribute{Key: "badgeAddress", Value: badgeData.Badge.Address},
+			utils.Attribute{Key: "badgeLavaChainId", Value: badgeData.Badge.LavaChainId},
+			utils.Attribute{Key: "badgeEpoch", Value: badgeData.Badge.Epoch},
+			utils.Attribute{Key: "relayAddress", Value: client},
+			utils.Attribute{Key: "relayLavaChainId", Value: relay.LavaChainId},
+			utils.Attribute{Key: "relayEpoch", Value: relay.Epoch},
+		)
+	}
+
+	badgeUsedCuKey := types.BadgeUsedCuKey(badgeData.Badge.ProjectSig, relay.Provider)
+	badgeUsedCuMapEntry, found := k.GetBadgeUsedCu(ctx, badgeUsedCuKey)
+	if !found {
+		// setting a timer with a callback to delete the badgeUsedCuEntry after badge.Epoch+blocksToSave (see keeper.go)
+		badgeUsedCuTimerExpiryBlock := k.BadgeUsedCuExpiry(ctx, badgeData.Badge)
+		if badgeUsedCuTimerExpiryBlock <= uint64(ctx.BlockHeight()) {
+			return 0, utils.LavaFormatWarning("badge rejected", fmt.Errorf("badge used CU entry validity expired"),
+				utils.Attribute{Key: "badgeUsedCuTimerExpiryBlock", Value: badgeUsedCuTimerExpiryBlock},
+				utils.Attribute{Key: "currentBlock", Value: uint64(ctx.BlockHeight())},
+			)
+		}
+		newTimerExpiry = badgeUsedCuTimerExpiryBlock
+		badgeUsedCuMapEntry = types.BadgeUsedCu{
+			UsedCu: 0,
+		}
+	}
+
+	// enforce badge CU overuse
+	if relay.CuSum+badgeUsedCuMapEntry.UsedCu > badgeData.Badge.CuAllocation {
+		return newTimerExpiry, utils.LavaFormatWarning("badge CU allocation exceeded", fmt.Errorf("could not update badge's used CU"),
+			utils.Attribute{Key: "relayCuSum", Value: relay.CuSum},
+			utils.Attribute{Key: "badgeCuLeft", Value: badgeData.Badge.CuAllocation - badgeUsedCuMapEntry.UsedCu},
+		)
+	}
+
+	return newTimerExpiry, nil
+}
+
+func (k Keeper) handleBadgeCu(ctx sdk.Context, badgeData BadgeData, provider string, relayCuSum uint64, newTimerExpiry uint64) {
+	badgeUsedCuKey := types.BadgeUsedCuKey(badgeData.Badge.ProjectSig, provider)
+	badgeUsedCuMapEntry, found := k.GetBadgeUsedCu(ctx, badgeUsedCuKey)
```

Comment:
we have it from the previous method, why not use it instead of getting it again?

Reply:
I'll skip that (as we talked)